File: Samples/Task_0 in CS/Player.cs
namespace Task_0
{
    internal class Player
    {
        private IWeaponStrategy _weaponStrategy;

        // Анализ проблемы: 
        // Нарушение открытости-закрытости
        // Для добавление нового оружия, нужно переписать код 
        // Длинная лестница if-else

        // Было так:
        //public void Attack(string weapon_type)
        //{
        //    if (weapon_type == "Sword")
        //        Console.WriteLine("Sword Attack!");

        //    if (weapon_type == "Bow")
        //        Console.WriteLine("Bow Attack!");

        //    if (weapon_type == "Magic")
        //        Console.WriteLine("Magic Attack!");
        //}

        // Стало: 
        public void SetWeaponStrategy(IWeaponStrategy weaponStrategy)
        {
            _weaponStrategy = weaponStrategy;
        }

        public void Attack()
        {
            if (_weaponStrategy == null)
            {
                Console.WriteLine("WeaponStrategy is NULL!");
                return;
            }

            _weaponStrategy.Attack();
        }
    }
}

File: Samples/Task_0 in CPP/Player.cpp
#include "Player.h"
#include <iostream>

// Áûëî
//void Player::Attack(string weapon_type)
//{
//	// Ïî÷åìó òàê ïëîõî?
//	if (weapon_type == "Sword")
//		cout << "Sword Attack!" << endl;
//
//	if (weapon_type == "Magic")
//		cout << "Magic Attack!" << endl;
//
//	if (weapon_type == "Bow")
//		cout << "Bow Attack!" << endl;
//}

// Ñòàëî:
void Player::SetWeapon(IWeaponStrategy* weaponStrategy)
{
	if (weaponStrategy)
		delete this->weaponStrategy;
	
	this->weaponStrategy = weaponStrategy;
}

void Player::Attack()
{
	if (weaponStrategy) {
		weaponStrategy->Attack();
	}
	else {
		cout << "Warning! Weapon Strategy not set!" << endl;
	}
}



